controlplane/internal/controllers/rke2controlplane_controller.go

```diff
@@ -340,34 +424,122 @@ func (r *RKE2ControlPlaneReconciler) reconcileNormal(ctx context.Context, cluste
+	// NOTE: Ignoring failures given that we are deleting
+	if _, err := r.reconcileControlPlaneConditions(ctx, controlPlane); err != nil {
+		logger.Info("failed to reconcile conditions", "error", err.Error())
+	}
+
+	// Aggregate the operational state of all the machines; while aggregating we are adding the
+	// source ref (reason@machine/name) so the problem can be easily tracked down to its source machine.
+	// However, during delete we are hiding the counter (1 of x) because it does not make sense given that
+	// all the machines are deleted in parallel.
+	conditions.SetAggregate(rcp, controlplanev1.MachinesReadyCondition, ownedMachines.ConditionGetters(), conditions.AddSourceRef(), conditions.WithStepCounterIf(false))
+
+	// Verify that only control plane machines remain
+	if len(allMachines) != len(ownedMachines) {
+		logger.Info("Waiting for worker nodes to be deleted first")
+		conditions.MarkFalse(rcp, controlplanev1.ResizedCondition, clusterv1.DeletingReason, clusterv1.ConditionSeverityInfo, "Waiting for worker nodes to be deleted first")
+		return ctrl.Result{RequeueAfter: deleteRequeueAfter}, nil
+	}
+
+	// Delete control plane machines in parallel
+	machinesToDelete := ownedMachines.Filter(collections.Not(collections.HasDeletionTimestamp))
+	var errs []error
+	for i := range machinesToDelete {
+		m := machinesToDelete[i]
+		logger := logger.WithValues("machine", m)
+		if err := r.Client.Delete(ctx, machinesToDelete[i]); err != nil && !apierrors.IsNotFound(err) {
+			logger.Error(err, "Failed to cleanup owned machine")
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) > 0 {
+		err := kerrors.NewAggregate(errs)
+		r.recorder.Eventf(rcp, corev1.EventTypeWarning, "FailedDelete",
+			"Failed to delete control plane Machines for cluster %s/%s control plane: %v", cluster.Namespace, cluster.Name, err)
+		return ctrl.Result{}, err
+	}
+	conditions.MarkFalse(rcp, controlplanev1.ResizedCondition, clusterv1.DeletingReason, clusterv1.ConditionSeverityInfo, "")
+	return ctrl.Result{RequeueAfter: deleteRequeueAfter}, nil
 }
 
+// 	// TODO: Improve this part once there are dependencies on the Control Plane Object!
+// 	var ok bool = true
+// 	for _, finalizer := range rcp.GetObjectMeta().GetFinalizers() {
+// 		if ok {
+// 			ok = controllerutil.RemoveFinalizer(rcp, finalizer)
+// 		}
+// 	}
+// 	if !ok {
+// 		logger.Info("unable to remove all finalizers")
+// 		err = fmt.Errorf("unable to remove all finalizers")
+// 		res = ctrl.Result{}
+// 		return
+// 	}
+
+// 	if err = r.Delete(ctx, rcp); err != nil {
+// 		res = ctrl.Result{RequeueAfter: 2 * time.Minute}
+// 		return
+// 	}
+
+// 	return ctrl.Result{}, nil
+// }
+
 func (r *RKE2ControlPlaneReconciler) reconcileKubeconfig(
 	ctx context.Context,
-	namespacedName types.NamespacedName,
+	clusterName client.ObjectKey,
 	endpoint clusterv1.APIEndpoint,
 	rcp *controlplanev1.RKE2ControlPlane) (ctrl.Result, error) {
 
+	if endpoint.IsZero() {
+		return ctrl.Result{}, nil
```

Comment:
Worth a high verbosity log that there is no endpoint yet

Reply:
Done